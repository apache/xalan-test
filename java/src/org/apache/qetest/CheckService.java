/*
 * The Apache Software License, Version 1.1
 *
 *
 * Copyright (c) 2000-2003 The Apache Software Foundation.  All rights 
 * reserved.
 *
 * Redistribution and use in source and binary forms, with or without
 * modification, are permitted provided that the following conditions
 * are met:
 *
 * 1. Redistributions of source code must retain the above copyright
 *    notice, this list of conditions and the following disclaimer. 
 *
 * 2. Redistributions in binary form must reproduce the above copyright
 *    notice, this list of conditions and the following disclaimer in
 *    the documentation and/or other materials provided with the
 *    distribution.
 *
 * 3. The end-user documentation included with the redistribution,
 *    if any, must include the following acknowledgment:  
 *       "This product includes software developed by the
 *        Apache Software Foundation (http://www.apache.org/)."
 *    Alternately, this acknowledgment may appear in the software itself,
 *    if and wherever such third-party acknowledgments normally appear.
 *
 * 4. The names "Xalan" and "Apache Software Foundation" must
 *    not be used to endorse or promote products derived from this
 *    software without prior written permission. For written 
 *    permission, please contact apache@apache.org.
 *
 * 5. Products derived from this software may not be called "Apache",
 *    nor may "Apache" appear in their name, without prior written
 *    permission of the Apache Software Foundation.
 *
 * THIS SOFTWARE IS PROVIDED ``AS IS'' AND ANY EXPRESSED OR IMPLIED
 * WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED WARRANTIES
 * OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE
 * DISCLAIMED.  IN NO EVENT SHALL THE APACHE SOFTWARE FOUNDATION OR
 * ITS CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
 * SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT
 * LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF
 * USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND
 * ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY,
 * OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT
 * OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF
 * SUCH DAMAGE.
 * ====================================================================
 *
 * This software consists of voluntary contributions made by many
 * individuals on behalf of the Apache Software Foundation and was
 * originally based on software copyright (c) 2000, Lotus
 * Development Corporation., http://www.lotus.com.  For more
 * information on the Apache Software Foundation, please see
 * <http://www.apache.org/>.
 */

package org.apache.qetest;

/**
 * Interface for 'check'ing (validating) equivalence of two items.
 *
 * Implementers provide their own algorithims for determining
 * equivalence, including custom algorithims for complex objects. 
 * CheckServices should log out both any pertinent information about 
 * the checking they've performed, then log out a checkPass or 
 * checkFail (or Ambg, etc.) record, as well as returning the 
 * appropriate result constant to the caller.
 *
 * The Configurable interface also allows callers to attempt to 
 * set attributes on either the CheckService or on any 
 * underlying validation algorithims.
 *
 * @author Shane_Curcuru@lotus.com
 * @version $Id$
 */
public interface CheckService extends Configurable
{

    /**
     * Compare two objects for equivalence, and return appropriate result.
     * Implementers should provide the details of their "equals"
     * algorithim in getDescription().  They must also call the 
     * appropriate checkPass()/checkFail()/etc. method on the 
     * supplied Logger.
     * 
     * Note that the order of actual, reference is usually 
     * important in determining the result.
     * <li>Typically:
     * <ul>any unexpected Exceptions thrown -> ERRR_RESULT</ul>
     * <ul>actual does not exist -> FAIL_RESULT</ul>
     * <ul>reference does not exist -> AMBG_RESULT</ul>
     * <ul>actual is equivalent to reference -> PASS_RESULT</ul>
     * <ul>actual is not equivalent to reference -> FAIL_RESULT</ul>
     * </li>
     *
     * @param logger to dump any output messages to
     * @param actual (current) Object to check
     * @param reference (gold, or expected) Object to check against
     * @param description of what you're checking
     * @param msg comment to log out with this test point
     * @return Logger.*_RESULT code denoting status; each method may 
     * define it's own meanings for pass, fail, ambiguous, etc.
     */
    public int check(Logger logger, Object actual,
                     Object reference, String msg);

    /**
     * Compare two objects for equivalence, and return appropriate result.
     *
     * @param logger to dump any output messages to
     * @param actual (current) Object to check
     * @param reference (gold, or expected) Object to check against
     * @param description of what you're checking
     * @param msg comment to log out with this test point
     * @param id ID tag to log out with this test point
     * @return Logger.*_RESULT code denoting status; each method may 
     * define it's own meanings for pass, fail, ambiguous, etc.
     */
    public int check(Logger logger, Object actual,
                     Object reference, String msg, String id);

    /**
     * Gets extended information about the last check call.
     * 
     * This is somewhat optional, and may be removed.  CheckServices 
     * should probably log out any additional info themselves to 
     * their logger before calling checkPass, etc., thus removing 
     * the need for callers to explicitly ask for this info.
     *
     * @return String describing any additional info about the last
     * two Objects that were checked, or null if none available
     */
    public String getExtendedInfo();

}  // end of class CheckService

