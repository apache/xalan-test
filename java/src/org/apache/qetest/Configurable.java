/*
 * The Apache Software License, Version 1.1
 *
 *
 * Copyright (c) 2001-2003 The Apache Software Foundation.  All rights 
 * reserved.
 *
 * Redistribution and use in source and binary forms, with or without
 * modification, are permitted provided that the following conditions
 * are met:
 *
 * 1. Redistributions of source code must retain the above copyright
 *    notice, this list of conditions and the following disclaimer. 
 *
 * 2. Redistributions in binary form must reproduce the above copyright
 *    notice, this list of conditions and the following disclaimer in
 *    the documentation and/or other materials provided with the
 *    distribution.
 *
 * 3. The end-user documentation included with the redistribution,
 *    if any, must include the following acknowledgment:  
 *       "This product includes software developed by the
 *        Apache Software Foundation (http://www.apache.org/)."
 *    Alternately, this acknowledgment may appear in the software itself,
 *    if and wherever such third-party acknowledgments normally appear.
 *
 * 4. The names "Xalan" and "Apache Software Foundation" must
 *    not be used to endorse or promote products derived from this
 *    software without prior written permission. For written 
 *    permission, please contact apache@apache.org.
 *
 * 5. Products derived from this software may not be called "Apache",
 *    nor may "Apache" appear in their name, without prior written
 *    permission of the Apache Software Foundation.
 *
 * THIS SOFTWARE IS PROVIDED ``AS IS'' AND ANY EXPRESSED OR IMPLIED
 * WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED WARRANTIES
 * OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE
 * DISCLAIMED.  IN NO EVENT SHALL THE APACHE SOFTWARE FOUNDATION OR
 * ITS CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
 * SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT
 * LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF
 * USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND
 * ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY,
 * OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT
 * OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF
 * SUCH DAMAGE.
 * ====================================================================
 *
 * This software consists of voluntary contributions made by many
 * individuals on behalf of the Apache Software Foundation and was
 * originally based on software copyright (c) 2000, Lotus
 * Development Corporation., http://www.lotus.com.  For more
 * information on the Apache Software Foundation, please see
 * <http://www.apache.org/>.
 */

package org.apache.qetest;
import java.util.Properties;

/**
 * Interface with basic configuration API's.  
 *
 * A simple interface allowing configuration of testing utilities; 
 * generally by passing generic information to the testing 
 * utility and having it pass it to any underlying objects.
 *
 * @author Shane_Curcuru@lotus.com
 * @version $Id$
 */
public interface Configurable
{
    /**
     * Allows the user to set specific attributes on the testing 
     * utility or it's underlying product object under test.
     * 
     * This method may or may not throw an IllegalArgumentException 
     * if an attribute that it does not recognize is passed in.  
     * This decision is up to the individual testing utility, and 
     * may include deciding to either trap underlying product 
     * exceptions for unknown attributes, or may re-throw them as 
     * IllegalArgumentExceptions.
     * 
     * Modeled after JAXP's various setAttribute()/setFeature() 
     * methods.
     * 
     * @param name The name of the attribute.
     * @param value The value of the attribute.
     * @throws IllegalArgumentException thrown if the underlying
     * implementation doesn't recognize the attribute and wants to 
     * inform the user of this fact.
     */
    public void setAttribute(String name, Object value)
        throws IllegalArgumentException;


    /**
     * Allows the user to set specific attributes on the testing 
     * utility or it's underlying product object under test.
     * 
     * This method should attempt to set any applicable attributes 
     * found in the given attrs onto itself, and will ignore any and 
     * all attributes it does not recognize.  It should never 
     * throw exceptions.  This method may overwrite any previous 
     * attributes that were set.  Currently since this takes a 
     * Properties block you may only be able to set objects that 
     * are Strings, although individual implementations may 
     * attempt to use Hashtable.get() on only the local part.
     * 
     * @param attrs Props of various name, value attrs.
     */
    public void applyAttributes(Properties attrs);


    /**
     * Allows the user to retrieve specific attributes on the testing 
     * utility or it's underlying product object under test.
     *
     * @param name The name of the attribute.
     * @return value The value of the attribute.
     * @throws IllegalArgumentException thrown if the underlying
     * implementation doesn't recognize the attribute and wants to 
     * inform the user of this fact.
     */
    public Object getAttribute(String name)
        throws IllegalArgumentException;


    /**
     * Description of what this testing utility does.  
     * 
     * @return String description of extension
     */
    public String getDescription();

}  // end of class Configurable

