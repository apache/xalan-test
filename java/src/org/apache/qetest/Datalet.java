/*
 * The Apache Software License, Version 1.1
 *
 *
 * Copyright (c) 2001-2003 The Apache Software Foundation.  All rights 
 * reserved.
 *
 * Redistribution and use in source and binary forms, with or without
 * modification, are permitted provided that the following conditions
 * are met:
 *
 * 1. Redistributions of source code must retain the above copyright
 *    notice, this list of conditions and the following disclaimer. 
 *
 * 2. Redistributions in binary form must reproduce the above copyright
 *    notice, this list of conditions and the following disclaimer in
 *    the documentation and/or other materials provided with the
 *    distribution.
 *
 * 3. The end-user documentation included with the redistribution,
 *    if any, must include the following acknowledgment:  
 *       "This product includes software developed by the
 *        Apache Software Foundation (http://www.apache.org/)."
 *    Alternately, this acknowledgment may appear in the software itself,
 *    if and wherever such third-party acknowledgments normally appear.
 *
 * 4. The names "Xalan" and "Apache Software Foundation" must
 *    not be used to endorse or promote products derived from this
 *    software without prior written permission. For written 
 *    permission, please contact apache@apache.org.
 *
 * 5. Products derived from this software may not be called "Apache",
 *    nor may "Apache" appear in their name, without prior written
 *    permission of the Apache Software Foundation.
 *
 * THIS SOFTWARE IS PROVIDED ``AS IS'' AND ANY EXPRESSED OR IMPLIED
 * WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED WARRANTIES
 * OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE
 * DISCLAIMED.  IN NO EVENT SHALL THE APACHE SOFTWARE FOUNDATION OR
 * ITS CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
 * SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT
 * LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF
 * USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND
 * ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY,
 * OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT
 * OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF
 * SUCH DAMAGE.
 * ====================================================================
 *
 * This software consists of voluntary contributions made by many
 * individuals on behalf of the Apache Software Foundation and was
 * originally based on software copyright (c) 2000, Lotus
 * Development Corporation., http://www.lotus.com.  For more
 * information on the Apache Software Foundation, please see
 * <http://www.apache.org/>.
 */

/*
 *
 * Datalet.java
 *
 */
package org.apache.qetest;

import java.util.Hashtable;

/**
 * Minimal interface defining a datalet, a single set of data 
 * for a simple test case.
 * <p>A Datalet defines a single group of data that a matching 
 * Testlet needs to execute it's simple test case.  Normally,
 * Testlets and Datalets are matched.</p>
 *
 * <p>This makes creating data-driven tests simpler, by separating 
 * the test algorithim from the definition of the test data.  Note 
 * that logging what happened during the test is already separated 
 * out into the Logger interface.</p>
 *
 * <p>Normally Datalets will simply be collections of public members
 * that can be written or read by anyone.  An enclosing test can 
 * simply create or load a number of Datalets and then pass them 
 * to a Testlet to execute the whole bunch of them.</p>
 *
 * <p>One way to look at a Datalet is basically an intelligent 
 * Properties / Hashtable / InputSource, designed for a 
 * specific set of data.  They can be loaded from a variety of 
 * inputs and will intelligently create the right kind of data 
 * that their corresponding Testlet is expecting.
 * For example, and DOM-based InputSource Datalet might be 
 * loaded from a number of Strings denoting filenames or URLs.
 * The Datalet would then know how to parse each of the files 
 * into a DOM, which a Testlet could then use directly.</p>
 *
 * <p>//@todo what metaphor is best? I.e. Should Datalets break OO
 * paradigms and just have public members, or should they work more 
 * like a Hashtable with some Properties-like features thrown in?
 * Or: what are the most important 
 * things to optimize: syntactic sugar-like simplicity for callers; 
 * more functionality for different datatypes; ease of use by 
 * inexperienced coders (perhaps testers who are just learning Java);
 * ease of use by experienced coders?</p>
 * 
 *
 * <p>//@todo Should we add a getParameterInfo() method?</p>
 * @author Shane_Curcuru@lotus.com
 * @version $Id$
 */
public interface Datalet
{

    /**
     * Accesor method for a brief description of this Datalet.  
     *
     * @return String describing the specific set of data 
     * this Datalet contains (can often be used as the description
     * of any check() calls made from the Testlet).
     */
    public abstract String getDescription();


    /**
     * Accesor method for a brief description of this Datalet.  
     * Datalets must have this as a read/write property, since many 
     * users will programmatically construct Datalets.
     *
     * @param s description to use for this Datalet.
     */
    public abstract void setDescription(String s);


    /**
     * Load fields of this Datalet from a Hashtable.
     * Note many datalets might take in a Properties block 
     * instead for simple String-valued data.
     *
     * @param Hashtable to load; Datalet should attempt to fill 
     * in as many member variables as possible from this, leaving 
     * any non-specified variables null or some default; behavior 
     * if null is passed is undefined
     */
    public abstract void load(Hashtable h);


    /**
     * Load fields of this Datalet from an array.
     * This allows most Datalets to be initialized from a main()
     * command line or the like.  For any String-valued types, or 
     * types that can easily be derived from a String, this makes 
     * running ad-hoc tests very easy.
     *
     * @param args array of Strings, as if from the command line;
     * Datalet should attempt to fill in as many member variables 
     * as possible from this, defining the order itself, leaving 
     * any non-specified variables null or some default; behavior 
     * if null is passed is undefined
     */
    public abstract void load(String[] args);

}  // end of class Datalet

