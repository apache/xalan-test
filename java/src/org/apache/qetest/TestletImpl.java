/*
 * The Apache Software License, Version 1.1
 *
 *
 * Copyright (c) 2001-2003 The Apache Software Foundation.  All rights 
 * reserved.
 *
 * Redistribution and use in source and binary forms, with or without
 * modification, are permitted provided that the following conditions
 * are met:
 *
 * 1. Redistributions of source code must retain the above copyright
 *    notice, this list of conditions and the following disclaimer. 
 *
 * 2. Redistributions in binary form must reproduce the above copyright
 *    notice, this list of conditions and the following disclaimer in
 *    the documentation and/or other materials provided with the
 *    distribution.
 *
 * 3. The end-user documentation included with the redistribution,
 *    if any, must include the following acknowledgment:  
 *       "This product includes software developed by the
 *        Apache Software Foundation (http://www.apache.org/)."
 *    Alternately, this acknowledgment may appear in the software itself,
 *    if and wherever such third-party acknowledgments normally appear.
 *
 * 4. The names "Xalan" and "Apache Software Foundation" must
 *    not be used to endorse or promote products derived from this
 *    software without prior written permission. For written 
 *    permission, please contact apache@apache.org.
 *
 * 5. Products derived from this software may not be called "Apache",
 *    nor may "Apache" appear in their name, without prior written
 *    permission of the Apache Software Foundation.
 *
 * THIS SOFTWARE IS PROVIDED ``AS IS'' AND ANY EXPRESSED OR IMPLIED
 * WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED WARRANTIES
 * OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE
 * DISCLAIMED.  IN NO EVENT SHALL THE APACHE SOFTWARE FOUNDATION OR
 * ITS CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
 * SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT
 * LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF
 * USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND
 * ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY,
 * OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT
 * OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF
 * SUCH DAMAGE.
 * ====================================================================
 *
 * This software consists of voluntary contributions made by many
 * individuals on behalf of the Apache Software Foundation and was
 * originally based on software copyright (c) 2000, Lotus
 * Development Corporation., http://www.lotus.com.  For more
 * information on the Apache Software Foundation, please see
 * <http://www.apache.org/>.
 */

/*
 *
 * TestletImpl.java
 *
 */
package org.apache.qetest;

import java.lang.reflect.Constructor;

/**
 * Simple implementation of a testlet, a sort of mini-test.
 * <p>A TestletImpl defines some common implementations that 
 * may be useful, including sample implementations that 
 * can be copied if you don't want to exend this class.</p>
 *
 * <p>The most useful implementation is of main(String[]), which 
 * allows a Testlet to be executed independently from the command 
 * line.  See the code comments for a way to get this behavior in 
 * your testlet without having to re-implement the whole main 
 * method - by just including a static{} initializer with the 
 * fully qualified classname of your class.</p>
 *
 * <b>Note:</b> Testlets based on this class are probably 
 * not threadsafe, and must be executed singly! 
 * See comments for thisClassName.
 * 
 * @author Shane_Curcuru@lotus.com
 * @version $Id$
 */
public class TestletImpl implements Testlet
{

    //-----------------------------------------------------
    //---- Implement Testlet interface methods
    //-----------------------------------------------------

    /**
     * Accesor method for a brief description of this test.  
     *
     * @return String "TestletImpl: default implementation, does nothing"
     */
    public String getDescription()
    {
        return "TestletImpl: default implementation, does nothing";
    }


    /**
     * Accesor methods for our Logger.
     *
     * @param l the Logger to have this test use for logging 
     * results; or null to use a default logger
     */
    public void setLogger(Logger l)
	{
        // if null, set a default one
        if (null == l)
            logger = getDefaultLogger();
        else
            logger = l;
	}


    /**
     * Accesor methods for our Logger.  
     *
     * @return Logger we tell all our secrets to.
     */
    public Logger getLogger()
	{
        return logger;
	}


    /**
     * Get a default Logger for use with this Testlet.  
     * Gets a default ConsoleLogger (only if a Logger isn't 
     * currently set!).  
     *
     * @return current logger; if null, then creates a 
     * Logger.DEFAULT_LOGGER and returns that; if it cannot
     * create one, throws a RuntimeException
     */
    public Logger getDefaultLogger()
    {
        if (logger != null)
            return logger;

        try
        {
            Class rClass = Class.forName(Logger.DEFAULT_LOGGER);
            return (Logger)rClass.newInstance();
        } 
        catch (Exception e)
        {
            // Must re-throw the exception, since returning 
            //  null or the like could lead to recursion
            e.printStackTrace();
            throw new RuntimeException(e.toString());
        }
    }


    /**
     * Return this TestletImpl's default Datalet.  
     * 
     * @return Datalet <code>defaultDatalet</code>.
     */
    public Datalet getDefaultDatalet()
	{
        return defaultDatalet;
	}


    /**
     * Run this TestletImpl: execute it's test and return.
     * This must (obviously) be overriden by subclasses.  Here, 
     * we simply log a message for debugging purposes.
     *
     * @param Datalet to use as data points for the test.
     */
    public void execute(Datalet datalet)
	{
        logger.logMsg(Logger.STATUSMSG, "TestletImpl.execute(" + datalet + ")");
	}


    //-----------------------------------------------------
    //---- Implement useful worker methods and main()
    //-----------------------------------------------------

    /**
     * Process default command line args.  
     * Provides simple usage functionality: given a first arg of 
     * -h, -H, -?, prints a usage statement based on getDescription()
     * and on getDefaultDatalet().getDescription()
     * 
     * @param args command line args from the JVM
     * @return true if we got and handled any default command line 
     * args (i.e. you can quit now); false otherwise (i.e. you 
     * should go ahead and execute)
     */
    protected boolean handledDefaultArgs(String[] args)
    {
        // We don't handle null or blank args
        if ((null == args) || (0 == args.length))
            return false;

        // Provide basic processing for help, usage cases
        if ("-h".equals(args[0])
            || "-H".equals(args[0])
            || "-?".equals(args[0])
           )
        {
            logger.logMsg(Logger.STATUSMSG, thisClassName + " usage:");
            logger.logMsg(Logger.STATUSMSG, "    Testlet: " + getDescription());
            logger.logMsg(Logger.STATUSMSG, "    Datalet: " + getDefaultDatalet().getDescription());
            return true;
        }

        // Otherwise, don't handle any other args
        return false;
    }


    /**
     * Default implementation for command line use.  
     * Note subclasses can easily get the functionality we provide 
     * here without having to copy this method merely by copying the 
     * static initalization block shown above, replacing the 
     * "thisClassName" with their own FQCN.
     *
     * This default implementation installs a default Logger, then 
     * checks for and handles a few default command line args, and 
     * then executes the Testlet.
     *
     * //@todo How do we easily specify alternate Datalets or 
     * data to load a Datalet from?  What about the Datalet that 
     * actually wants the first arg to be -h?
     *
     * @param args command line args from the JVM
     */
    public static void main(String[] args)
    {
        if (true)
            System.out.println("TestletImpl.main");
        TestletImpl t = null;
        try
        {
            // Create an instance of the specific class at runtime
            //  This relies on subclasses to reset 'thisClassName'
            //  in their own static[] initialization block!
            t = (TestletImpl)Class.forName(thisClassName).newInstance();

            // Set a default logger automatically
            t.setLogger(t.getDefaultLogger());

            // Process default -h, etc. args
            if (!t.handledDefaultArgs(args))
            {
                if (args.length > 0)
                {
                    // If we do have any args, then attempt to 
                    //  load the correct-typed Datalet from the args
                    Class dataletClass = t.getDefaultDatalet().getClass();
                    t.logger.logMsg(Logger.TRACEMSG, "Loading Datalet " 
                                  +  dataletClass.getName() + " from args");
                    Datalet d = (Datalet)dataletClass.newInstance();
                    d.load(args);
                    t.execute(d);
                }
                else
                {
                    // Otherwise, use the defaultDatalet for that Testlet
                    t.logger.logMsg(Logger.TRACEMSG, "Using defaultDatalet");
                    t.execute(t.getDefaultDatalet());
                }
            }
        }
        catch (Exception e)
        {
            if ((null != t) && (null != t.logger))
            {
                // Use the logger which is (hopefully) OK
                t.logger.checkErr("TestletImpl threw: " + e.toString());
                java.io.StringWriter sw = new java.io.StringWriter();
                java.io.PrintWriter pw = new java.io.PrintWriter(sw);
                e.printStackTrace(pw);
                t.logger.logArbitrary(Logger.ERRORMSG, sw.toString());
            }
            else
            {
                // Otherwise, just dump to System.err
                System.err.println("TestletImpl threw: " + e.toString());
                e.printStackTrace();
            }
        }
    }


    /**
     * The FQCN of the current class.  
     * See comments in main() and in the static initializer.
     * <b>Note:</b> Testlets based on this class are probably 
     * not threadsafe, and must be executed singly! 
     */
    protected static String thisClassName = null;


    /**
     * A static initializer setting the value of thisClassName.  
     * Subclasses must copy this static block, and replace the 
     * name of "...TestletImpl" with their own name.  Then, when 
     * a user executes the subclassed Testlet on the command line 
     * or calls it's main() method, the correct thing will happen.
     */
    static { thisClassName = "org.apache.qetest.TestletImpl"; }


    /**
     * Our Logger, who we tell all our secrets to.  
     */
    protected Logger logger = null;


    /**
     * Our deafult Datalet: in this case, not very interesting.  
     * Provide a default 'null' datalet so unsuspecting callers 
     * don't get NullPointerExceptions.
     */
    protected Datalet defaultDatalet = new NullDatalet();

}  // end of class TestletImpl
