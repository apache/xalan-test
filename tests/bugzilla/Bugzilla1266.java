/*
 * Licensed to the Apache Software Foundation (ASF) under one
 * or more contributor license agreements. See the NOTICE file
 * distributed with this work for additional information
 * regarding copyright ownership. The ASF licenses this file
 * to you under the Apache License, Version 2.0 (the  "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
// Common Qetest / Xalan testing imports
import org.apache.qetest.Datalet;
import org.apache.qetest.Logger;
import org.apache.qetest.TestletImpl;

// REPLACE_imports needed for reproducing the bug
import org.apache.qetest.*;
import org.apache.qetest.trax.*;
import org.apache.qetest.xsl.*;

// Import all relevant TRAX packages
import javax.xml.transform.*;
import javax.xml.transform.dom.*;
import javax.xml.transform.sax.*;
import javax.xml.transform.stream.*;
import javax.xml.parsers.DocumentBuilder;
import javax.xml.parsers.DocumentBuilderFactory;
import javax.xml.parsers.SAXParserFactory;

// Needed SAX, DOM, JAXP classes
import org.xml.sax.InputSource;
import org.xml.sax.XMLReader;
import org.w3c.dom.Node;

// java classes
import java.io.File;
import java.io.FilenameFilter;
import java.util.Properties;

// jkesselm May2023: CONFIRMED that console shows
//   Warning:  The encoding 'illegal-encoding-value' is not supported by the Java runtime.
//   Warning: encoding "illegal-encoding-value" not supported, using UTF-8
// but LoggingErrorHandler is not invoked.
//
// This appears because at the time serializer.ToXMLStream.setProp() is invoked
// for this value, the object's SerializerBase.m_transformer field has not been
// initialized, preventing setProp() from being able to access the errHandler.
//
// To resolve this, I think we would want to have
// SerializerFactory.getSerializer() call ser.setTransformer() before
// ser.setOutputFormat. Unfortunately, as currently coded, the Transformer
// is not being passed in as an argument to SerializerFactory.
//
// Least-impact change might be to add a second SerializerFactory.getSerializer()
// method which takes Transformer as a second argument, and alter 
// TransformerImpl.createSerializationHandler() to call that entry point,
// passing in itself ("this").
//
// RECOMMENDATION: REVIEW THAT PROPOSAL.

/**
 * Testlet for reproducing Bugzilla reported bugs.
 *
 * @author shane_curcuru@lotus.com
 * @author wjboukni@eos.ncsu.edu
 */
public class Bugzilla1266 extends TestletImpl
{
    // Initialize our classname for TestletImpl's main() method - must be updated!
    static { thisClassName = "Bugzilla1266"; }

    /**
     * Write Minimal code to reproduce your Bugzilla bug report.
     * Many Bugzilla tests won't bother with a datalet; they'll 
     * just have the data to reproduce the bug encoded by default.
     * @param d (optional) Datalet to use as data point for the test.
     */
    public void execute(Datalet d)
    {
        // Use logger.logMsg(...) instead of System.out.println(...)
        logger.logMsg(Logger.STATUSMSG, "Reproducing Bugzilla#1266");
        LoggingErrorListener loggingErrorListener = new LoggingErrorListener(logger);
        loggingErrorListener.setThrowWhen(LoggingErrorListener.THROW_NEVER);
        logger.logMsg(Logger.STATUSMSG, "loggingErrorListener originally setup:" + loggingErrorListener.getQuickCounters());

        TransformerFactory factory = null;
        Templates templates = null;
        Transformer transformer = null;
        try
            {
                factory = TransformerFactory.newInstance();
                logger.logMsg(Logger.STATUSMSG, "About to factory.newTemplates(" + QetestUtils.filenameToURL("identity.xsl") + ")");
                templates = factory.newTemplates(new StreamSource(QetestUtils.filenameToURL("identity.xsl")));
                transformer = templates.newTransformer();

                // Set the errorListener and validate it
                transformer.setErrorListener(loggingErrorListener);
                if (transformer.getErrorListener() == loggingErrorListener)
                    logger.checkPass("set/getErrorListener on transformer");
                else {
                    logger.checkFail("set/getErrorListener on transformer");
		    throw new Exception("failed set/getErrorListener on transformer");
                }

                logger.logMsg(Logger.STATUSMSG, "Reproduce Bugzilla1266 - warning due to bad output props not propagated");
                logger.logMsg(Logger.STATUSMSG, "transformer.setOutputProperty(encoding, illegal-encoding-value)");
                transformer.setOutputProperty("encoding", "illegal-encoding-value");

                logger.logMsg(Logger.STATUSMSG, "about to transform(...)");
                transformer.transform(new StreamSource(QetestUtils.filenameToURL("identity.xml")), 
                                      new StreamResult("Bugzilla1266.out"));
                logger.logMsg(Logger.STATUSMSG, "after transform(...)");
                logger.logMsg(Logger.STATUSMSG, "loggingErrorListener after transform:" + loggingErrorListener.getQuickCounters());

                // Validate that one warning (about illegal-encoding-value) should have been reported
                int[] errCtr = loggingErrorListener.getCounters();
                            
                if (errCtr[LoggingErrorListener.TYPE_WARNING] > 0)
                    logger.checkPass("At least one Warning listned to for illegal-encoding-value");
                else {
                    logger.checkFail("No Warning reported by listener for illegal-encoding-value");
		    throw new Exception("No Warning reported by listener for illegal-encoding-value");
                }
                
                // Validate the actual output file as well: in this case, 
                //  the stylesheet should still work
                CheckService fileChecker = new XHTFileCheckService();
                fileChecker.check(logger, 
                                  new File("Bugzilla1266.out"), 
                                  new File("identity.xml"), 
                                  "transform of good xsl w/bad output props into: " + "Bugzilla1266.out");
            
            }
        catch (Throwable t)
            {
                logger.checkFail("Bugzilla1266 unexpectedly threw: " + t.toString());
                logger.logThrowable(Logger.ERRORMSG, t, "Bugzilla1266 unexpectedly threw");
                
            }

    }

    /**
     * <a href="http://nagoya.apache.org/bugzilla/show_bug.cgi?id=1266">
     * Link to Bugzilla report</a>
     * @return Warning Event not being fired from Transformer when using invalid Encoding String.
     */
    public String getDescription()
    {
        return "Warning Event not being fired from Transformer when using invalid Encoding String";
    }

}  // end of class Bugzilla1266

