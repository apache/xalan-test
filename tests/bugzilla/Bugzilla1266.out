<?xml version="1.0" encoding="UTF-8"?><doc>
  <title name="title-name-attr">title-content</title>
  <list>
    <item version="1">Xalan-J 1.x</item>
    <item version="2">Xalan-J 2.x</item>
    <item version="1">Xalan-C 1.x</item>
    <list>
      <item>Xalan documentation</item>
      <item>Xalan tests</item>
    </list>
  </list>
</doc>